README.md

```diff
@@ -0,0 +1,396 @@
+
+It is also a chance to try out some "interesting" ideas (such as config reducers) that would be much more difficult to sketch out in a real Agent.
+Maybe some of these are bad ideas, maybe they are things we can do later, or maybe we want to do them first thing.
+
+You can see the set of components in [`COMPONENTS.md`](./COMPONENTS.md).
+
+# Building
+
+To build, run `inv build`.
+This will build all binaries in the root of the repository.
+
+To regenerate files in the repo, run `inv generate`.
+
+# Component Guidelines
+
+What follows are draft guidelines for writing components.
+This repository follows these guidelines.
+
+If these guidelines are adopted, then the expectation is that all components would follow them.
+This reduces cognitive load when using components -- no need to remember one component's peculiarities.
+It also allows Agent-wide changes, where we make the same formulaic change to each component.
+If a situation arises that contradicts the guidelines, then we can update the guidelines (and change all affected components).
+
+## Prerequisites
+
+You should be familiar with [`fx`](https://pkg.go.dev/go.uber.org/fx), the tool used for dependency injection.
+
+## Component Definition
+
+A component is defined in a dedicated package under `comp/`, with the following defined in `component.go`:
+
+ * Extensive package-level documentation.
+   This should define, as precisely as possible, the behavior of the component, acting as a contract on which users of the component may depend.
+
+   The documentation (both package-level and method-level) should also include:
+
+   * Precise information about which interface methods can be called during the setup phase, and which must only be called after the component is started.
+   * Precise information about data ownership of passed values and returned values (does the method modify its arguments?  can the caller modify a returned slice or map?).
+   * Precise information about goroutines and blocking (does the method block? is a callback invoked in a dedicated goroutine? what happens if a channel is full?)
+
+ * `pkg.Component` -- the type implemented by the component.
+   This can be an empty interface, but is the type by which other components will find this one via `fx`.
+   It should have a formulaic doc string like `// Component is the component type.`, deferring documentation to the package docs.
+   All interface methods should be exported and thoroughly documented.
+
+ * `pkg.Module` -- an `fx.Option` that can be included in an `fx.App` to make this component available.
+   To improve logging, use `fx.Module(comppath, ..)`, where `comppath` is the root-relative package path, e.g., `"comp/util/log"`.
+   It should have a formulaic doc string like `// Module defines the fx options for this component.`
+
+Components should not be nested; that is, no component's Go path should be a prefix of another component's Go path.
+
+### Implementation
```

Comment:
I'm not too happy with this arrangement, either.  I'd like to "bundle" components together so that we don't need to maintain long lists of components all over the place.  I want to experiment with this a bit more.  I'd love others' ideas, too!

Reply:
I'm not sure internals should be exposed as components, at least "public" components.
In my understanding, in the end, each component or `InputType/ParamType` will need to be registered by the app entrypoint (directly on indirectly).

If `Dogstatsd` relies on something internal that is not shared with any other component, it may not make sense to put it as a component. If we still want to, to ease testing or whatever, the chosen `Dogstatsd` component should register them and I don't think it would a good practise for them to depend on components that are not internal or in the `ModuleParams` of `Dogstatsd` itself, to make the dependencies explicit.

Worst case, the developper did not pay attention and the binary will not run, which is easy to catch.

Reply:
I'm not sure what you mean by "public" component.  Fx doesn't provide a way to "hide" components, and those components would need to have dependencies outside of the bundle, too.  I think having a holistic view of "what's running" in the agent is actually helpful, even if some of those are named "comp/foo/internal/some/specific/thing".  Note that the type `comp/foo/internal/some/specific/thing.Component` isn't accessible outside of `comp/foo`, so components in other bundles can't depend on it - perhaps that achieves the "non-public" property you're looking for?

As for registration -- yes, that's what I'm working toward with bundles, but it's definitely not there yet.

Reply:
Also, thanks for the point about "easy to catch" -- unit tests are going to have a hard time catching incorrect combinations of components (since the units being tested are components or smaller).  So the more errors we can catch either at app build time or at invocation time, the better (assuming apps are invoked in CI).

Reply:
What I meant by "public" is components that **need** to be registered by the top-level `cmd`. Like I know I need to register `Dogstatsd` for my agent (call to `New` or `Provide`), but do I need to register `comp/foo/internal/some/specific/thing.Component` myself or is it taken care of by the `Dogstatsd` component/bundle?

To me, if I need to register it at top level, it's a "public" component/bundle and it means that I have control over some options for this component or the actual implementation of this component.

And indeed if we want to have an internal component for whatever reasons, at `comp/foo/internal/some/specific/thing.Component`, the top-level component must ensure that it's always provided for.

Reply:
Ah, I see what you mean.  The answer to "what is public" would be really easy if we weren't worried about binary size, but since we are I think it's going to be a little bit ad-hoc.  For example, with the logs-agent, we'll have to document that by default no launchers are included, and the app must include the desired launchers.  My hunch is that we can concentrate this "ad hoc" stuff into bundle documentation, so basically each bundle's got some special rules to follow but at least there are on the order of a dozen bundles so it's not a huge amount to remember (or look up).